### PR TITLE
[SPARK-41008][MLLIB] Dedup isotonic regression duplicate features

### PR DESCRIPTION
**Pull request by ahmed-mahran**
### What changes were proposed in this pull request?

Adding a pre-processing step to isotonic regression in mllib to handle duplicate features. This is to match `sklearn` implementation. Input points of duplicate feature values are aggregated into a single point using as label the weighted average of the labels of the points with duplicate feature values. All points for a unique feature values are aggregated as:
 - Aggregated label is the weighted average of all labels
 - Aggregated feature is the weighted average of all equal features. It is possible that feature values to be equal up to a resolution due to representation errors, since we cannot know which feature value to use in that case, we compute the weighted average of the features. Ideally, all feature values will be equal and the weighted average is just the value at any point.
 - Aggregated weight is the sum of all weights

### Why are the changes needed?

As per discussion on ticket [[SPARK-41008]](https://issues.apache.org/jira/browse/SPARK-41008), it is a bug and results should match `sklearn`.

### Does this PR introduce _any_ user-facing change?

There are no changes to the API, documentation or error messages. However, the user should expect results to change.

### How was this patch tested?

Existing test cases for duplicate features failed. These tests were adjusted accordingly. Also, new tests are added.

Here is a python snippet that can be used to verify the results:

```python
from sklearn.isotonic import IsotonicRegression

def test(x, y, x_test, isotonic=True):
    ir = IsotonicRegression(out_of_bounds='clip', increasing=isotonic).fit(x, y)
    y_test = ir.predict(x_test)
    
    def print_array(label, a):
        print(f"{label}: [{', '.join([str(i) for i in a])}]")

    print_array("boundaries", ir.X_thresholds_)
    print_array("predictions", ir.y_thresholds_)
    print_array("y_test", y_test)

test(
    x = [0.6, 0.6, 0.333, 0.333, 0.333, 0.20, 0.20, 0.20, 0.20],
    y = [1, 0, 0, 1, 0, 1, 0, 0, 0],
    x_test = [0.6, 0.6, 0.333, 0.333, 0.333, 0.20, 0.20, 0.20, 0.20]
)
```

@srowen @zapletal-martin

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala [hunk 1]
```diff
@@ -307,6 +307,99 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
...
+    if (cleanInput.isEmpty) Array.empty
+    else if (cleanInput.length <= 1) cleanInput
+    else {
+      // whether or not two double features are equal up to a precision
+      @inline def areEqual(a: Double, b: Double): Boolean = Precision.equals(a, b)
+
+      // Utility object, holds a buffer of all points with unique features so far, and performs
+      // weighted sum accumulation of points. Hides these details for better readability of the
+      // main algorithm.
+      object PointsAccumulator {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:348]: Maybe make this a top-level object in this file at least, if not a separate file

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala [hunk 2]
```diff
@@ -307,6 +307,99 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
...
+      input: Array[(Double, Double, Double)]): Array[(Double, Double, Double)] = {
+
+    val cleanInput = input.filter { case (y, x, weight) =>
+      require(
+        weight >= 0.0,
+        s"Negative weight at point ($y, $x, $weight). Weights must be non-negative")
+      weight > 0
+    }
+
+    if (cleanInput.isEmpty) Array.empty
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:339]: Expand the if-else so bodies are on new lines

File: mllib/src/test/scala/org/apache/spark/mllib/regression/IsotonicRegressionSuite.scala [hunk 3]
```diff
@@ -194,40 +241,46 @@ class IsotonicRegressionSuite extends SparkFunSuite with MLlibTestSparkContext w
     assert(model.predict(0.5) === 1.5)
     assert(model.predict(0.75) === 1.75)
     assert(model.predict(1) === 2)
-    assert(model.predict(2) === 10d/3)
-    assert(model.predict(9) === 10d/3)
+    assert(model.predict(2) === 10d / 3)
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/regression/IsotonicRegressionSuite.scala:244]: You're welcome to just write 10.0 / 3 here too

**Review by srowen — COMMENTED** [review]
Looking good, if tests pass. Is it fair to say this keeps existing tests and adds new ones? my only concern is about introducing new unintended behavior changes, but if it's passing all the previous tests, that's good enough for purposes here

**Comment by ahmed-mahran** [comment]
I don't remember whether I've checked all tests against sklearn. I'll double check anyways.

**Comment by ahmed-mahran** [comment]
Double-checking on this branch, all tests produce same results as sklearn.

On master, these two (already existing) tests produce different results than sklearn:
- [`test("isotonic regression prediction with duplicate features")`](https://github.com/apache/spark/pull/38966/files#diff-6eae24fe4c455659652a83f423f0df48ad994237b3350f18953c39258ae1ab86R248-R261)
- [`test("antitonic regression prediction with duplicate features")`](https://github.com/apache/spark/pull/38966/files#diff-6eae24fe4c455659652a83f423f0df48ad994237b3350f18953c39258ae1ab86R263-R276)

**Comment by srowen** [comment]
Ok if it was already varying from sklearn, and isn't the same issue as you're fixing, leave it for now

**Comment by ahmed-mahran** [comment]
Aha, I see your point! It was varying because of the issue I'm fixing, which is about handling duplicate features.

**Comment by ahmed-mahran** [comment]
... So, if we run this new code on existing (old) master's suite, those two tests will fail.

**Comment by srowen** [comment]
Ah OK, hm, that sounds like a regression related to this change then and probably needs to be fixed (or decide the old test was wrong)

**Comment by ahmed-mahran** [comment]
Regardless from this change, these tests' assumptions contradicts with what this ticket is about. See the tests' names at least: the tests are about handling duplicate features. My guess is that these were introduced to cover corner cases of lex sorting and repartitioning. However, we know that original implementation doesn't actually handle cosolidation of duplicate features. Hence, these tests by definition, are not valid.

**Comment by srowen** [comment]
OK so the test result is different, but maybe wasn't correct in the first place. But it doesn't match sklearn now, and did before (?) I'm still a little concerned that this changes behavior in unintended ways, but maybe you have a better judgment about whether this is still mostly an improvement in the result

**Comment by srowen** [comment]
Merged to master

**Comment by ahmed-mahran** [comment]
So, it is true: [scikit-learn has changed behavior](https://github.com/scikit-learn/scikit-learn/pull/4302) just after [spark introduced isotonic regression](https://github.com/apache/spark/pull/3519). I've tested different scikit-learn versions (namely: 0.13.1, 0.14.1, 0.15.0, 0.15.2) since isotonic regression was first introduced until the duplicates PR applied, and interestingly none has matched master's spark.

```python
from sklearn.isotonic import IsotonicRegression

IsotonicRegression(out_of_bounds='clip', increasing=isotonic).fit(x, y, sample_weight=weights).predict(x_test) # >= 0.15
IsotonicRegression(increasing=isotonic).fit(x, y, sample_weight=weights).predict(x_test) # = 0.14
IsotonicRegression().fit(x, y, weight=weights).predict(x_test) # = 0.13

# test("isotonic regression prediction with duplicate features")
x, y, weights, x_test, isotonic = [1, 1, 2, 2, 3, 3], [2, 1, 4, 2, 6, 5], [1, 1, 1, 1, 1, 1], [0, 1.5, 2.5, 4], True

# spark  : array([ 1 ,  2,  4.5,  6 ])
# v0.13.1: error
# v0.14.1: error
# v0.15.0: array([ nan,  2. ,  4.5,  5. ])
# v0.15.2: array([ 2. ,  2. ,  4.5,  6. ])
# v0.16.0: array([ 1.5 ,  2.25,  4.25,  5.5 ])

# test("antitonic regression prediction with duplicate features")
x, y, weights, x_test, isotonic = [1, 1, 2, 2, 3, 3], [5, 6, 2, 4, 1, 2], [1, 1, 1, 1, 1, 1], [0, 1.5, 2.5, 4], False

# spark  : array([ 6 ,  4.5,  2,  1 ])
# v0.13.1: error
# v0.14.1: error
# v0.15.0: array([  nan,  4.25,  2.25,  1.5 ])
# v0.15.2: array([ 5.5 ,  4.25,  2.25,  1.5 ])
# v0.16.0: array([ 5.5 ,  4.25,  2.25,  1.5 ])
```

[This R package](https://cran.r-project.org/web/packages/isotone/vignettes/isotone.pdf) suggests that there are three methods for breaking ties: primary, secondary and tertiary. Spark and scikit-learn originally implemented "primary". Then scikit-learn replaced it with "secondary" which is implemented in this PR.

**Comment by ahmed-mahran** [comment]
I think we need to follow up with documentation updates https://spark.apache.org/docs/latest/mllib-isotonic-regression.html#:~:text=If%20the%20prediction,point%20are%20used.

**Comment by srowen** [comment]
OK, as long as it's logical and consistent, and more logical/consistent than before, I think it's OK. Feel free to update docs as you see fit

**Comment by ahmed-mahran** [comment]
I'll follow up with PR for doc and removing the redundant sort key

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala [hunk 4]
```diff
@@ -434,12 +485,56 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
...
       .values
+      // Lexicographically sort points by features then labels.
       .mapPartitions(p => Iterator(p.toArray.sortBy(x => (x._2, x._1))))
+      // Aggregate points with equal features into a single point.
+      .map(makeUnique)
       .flatMap(poolAdjacentViolators)
       .collect()
-      .sortBy(x => (x._2, x._1)) // Sort again because collect() doesn't promise ordering.
+      // Sort again because collect() doesn't promise ordering.
+      .sortBy(x => (x._2, x._1))
```
[the remark this answers is not in the document]
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:498]: You can open up a follow-up PR on the same JIRA if there are minor additional improvements or docs to suggest
